### PR TITLE
remove threadsafe ptree json_parser

### LINKED ISSUE #1474: actually header-only ptree
so in #1465 we mentioned all of the dependencies that valhalla has. we are specifically trying to remove the ones that require harder to get third party libraries. ptree was one of these. its a header only library but if you use the json parser, this makes use of boost::spirit which isnt threadsafe. to make it threadsafe you have to use boost::thread, which isnt header-only. whew...

so we have a lot of ptree use in valhalla (which we started with rapidjson...). instead of rewriting everything in rapidjson (a lot of work but the eventual goal). what we can do is simply replace the boost json_parser with a call to rapidjson to parse the json, and a function to convert the rapidjson document into a ptree.

this will allow us to quickly remove the boost thread dependency and use ptree in header-only mode until we can spend the time to fully remove it and use rapidjson throughout.

### PR DESCRIPTION
**Pull request by kevinkreiser**
fixes #1474 

this is the first step in no longer using ptree for json. currently we have a bunch of different ways we interact with json. we use rapidjson for request parsing (where we used to use ptree) and we use ptree for configs and for loading language files and we use another thing for serializing json (also a dumb idea in hindsight). 

because our systems can be multithreaded we needed to make use of threadsafe mode for ptree. under the hood this requires boost::spirit and boost::thread. other than that ptree is a header-only library. we would like to use it in this way for the time being just because there so much code to change to switch to rapidjson.

so to make our build have less dependencies ive removed the property_tree::json_parser and replaced it with a method that uses rapidjson to do the parsing of the json and converts the rapidjson object into a ptree. eventually we want to go to full on rapidjson but it will take time and lots of code changes.

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -37,10 +37,6 @@ if(CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
   list(APPEND boost_components zlib)
 endif()
 find_package(Boost 1.51 REQUIRED COMPONENTS ${boost_components})
-if(${Boost_MAJOR_VERSION}.${Boost_MINOR_VERSION} VERSION_LESS 1.59)
```
Comment (kevinkreiser) [inline src/CMakeLists.txt]: we dont need threadsafe ptree anymore and so we dont need boost::thread

File: test/json.cc [hunk 2]
```diff
@@ -64,28 +64,9 @@ void TestJsonSerialize() {
...
-    ans_set.insert(line);
-  if (res_set.size() != ans_set.size())
-    throw std::logic_error("Wrong json!");
-
-  auto res_itr = res_set.cbegin(), ans_itr = ans_set.cbegin();
-  for (; res_itr != res_set.cend() && ans_itr != ans_set.cend(); ++res_itr, ++ans_itr)
-    if (*res_itr != *ans_itr)
-      throw std::runtime_error("Wrong json!");
+  rapidjson::Document res, ans;
+  if (res != ans)
```
Comment (kevinkreiser) [inline test/json.cc]: testing is easier with rapidjson

File: valhalla/baldr/rapidjson_utils.h [hunk 3]
```diff
@@ -148,6 +153,74 @@ inline boost::optional<rapidjson::Value&> get_child_optional(V&& v, const char*
...
+    add_value(e.value, pt.add_child(e.name.GetString(), Ptree{}));
+}
+
+template <class Ptree> void add_array(const GenericArray<true, Value::ValueType>& a, Ptree& pt) {
+  for (const auto& e : a)
+    add_value(e, pt.push_back(std::make_pair("", Ptree{}))->second);
+}
+
+template <class Ptree>
+void read_json(std::basic_istream<typename Ptree::key_type::value_type>& stream, Ptree& pt) {
```
Comment (kevinkreiser) [inline valhalla/baldr/rapidjson_utils.h]: this is the important bit. it has the same signature that is found in the `property_tree::json_parser` and so once i implemented it i didnt really have to change anything but the namespace and include in all the things that used it. this is basically the start of the recursion of making a ptree

File: valhalla/baldr/rapidjson_utils.h [hunk 4]
```diff
@@ -148,6 +153,74 @@ inline boost::optional<rapidjson::Value&> get_child_optional(V&& v, const char*
...
 
+template <class Ptree> void add_value(const Value& v, Ptree& pt) {
+  switch (v.GetType()) {
+    case kObjectType:
+      add_object(v.GetObject(), pt);
+      break;
+    case kArrayType:
+      add_array(v.GetArray(), pt);
+      break;
+    case kNullType:
```
Comment (kevinkreiser) [inline valhalla/baldr/rapidjson_utils.h]: any case from here down is a base case that stops the recursion. the calls to add_object and add_array deepen the call stack

File: src/CMakeLists.txt (hunk 1 above)
Comment (dnesbitt61) [reply src/CMakeLists.txt]: yes!

**Pull request merged**